include/aws/common/encoding.h

```diff
@@ -0,0 +1,96 @@
+#ifndef AWS_COMMON_ENCODING_H
+#define AWS_COMMON_ENCODING_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * Base 16 (hex) encodes the contents of to_encode and stores the result in output.
+ * If output is NULL, output_size will be set to what the output_size should be.
+ */
+AWS_COMMON_API int aws_hex_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size);
+
+/*
+ * Base 16 (hex) decodes the contents of to_decode and stores the result in output.
+ * If output is NULL, output_size will be set to what the output_size should be.
+ */
+AWS_COMMON_API int aws_hex_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size);
+
+/*
+ * Base 64 encodes the contents of to_encode and stores the result in output.
+ * If output is NULL, output_size will be set to what the output_size should be.
+ */
+AWS_COMMON_API int aws_base64_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size);
+
+/*
+ * Base 64 decodes the contents of to_decode and stores the result in output.
+ * If output is NULL, output_size will be set to what the output_size should be.
+ */
+AWS_COMMON_API int aws_base64_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size);
+
+/*Add a 64 bit unsigned integer to the buffer, ensuring network - byte order
+ * return the new position in the buffer for the next operation.
+ */
+AWS_COMMON_API uint8_t *aws_add_uint64_to_buffer(uint8_t *buffer, uint64_t value);
```

Comment:
I would recommend a simpler API:

```
/*
 * Write a 64-bit value to *buf in network byte order. buf need not be aligned.
 */
void aws_write_u64(void *buf, uint64_t value);

/*
 * Reads a 64-bit value from *buf in network byte order. buf need not be aligned.
 */
uint64_t aws_read_u64(const void *buf);
```

There's not a lot of value add for returning the adjusted pointer unless you're also doing bounds checks - in which case you might be able to add some spectre mitigations here. I have some prototype code along these lines I can share.

This is also another place where I would strongly recommend writing these as static inline functions in a header file. An unaligned 64-bit big-endian read is two instructions on x86_64, whereas a function call implies a lot of overhead (register spills, etc).

Reply:
If we do static inline, we can never expose this interface in the form of bindings for other languages.

I think I agree on the first statement.

Reply:
That's a good point. We can however arrange to `static inline` but also emit an actual implementation, even without the C99 `extern inline`. For example, we could have, instead of `static inline`:

```
#ifndef AWS_INLINE
#define AWS_INLINE static inline
#endif
```

Then, in an `inlines.c`:

```
#define AWS_INLINE
#include <aws/common/encoding.h>
#include <aws/common/list.h>
// ...
#endif
```

C code will use the `static inline` definition; however, an implementation of the function will still be in the library for use by FFIs.

Reply:
on windows it needs to be dll_export etc... to get exported. I think this is the right direction, I just need to think on what the macro trickery would be. Largely, I want to avoid different assemblies based on the consumer.